# 面向过程编程
## 函数

In [1]:
def my_abs(x):
    if x >= 0:
        return x
    else:
        return -x
print(my_abs(-99))

99


如果想定义一个什么事也不做的空函数，可以用pass语句：
pass语句什么都不做，那有什么用？实际上pass可以用来作为占位符，比如现在还没想好怎么写函数的代码，就可以先放一个pass，让代码能运行起来。

In [ ]:
def nop():
    pass

调用函数时，如果参数个数不对，Python解释器会自动检查出来，并抛出TypeError：

In [2]:
 my_abs(1, 2)

TypeError: my_abs() takes 1 positional argument but 2 were given

但是如果参数类型不对，Python解释器就无法帮我们检查。试试my_abs和内置函数abs的差别：

In [3]:
my_abs('A')
abs('A')

TypeError: '>=' not supported between instances of 'str' and 'int'

当传入了不恰当的参数时，内置函数abs会检查出参数错误，而我们定义的my_abs没有参数检查，会导致if语句出错，出错信息和abs不一样。所以，这个函数定义不够完善。
让我们修改一下my_abs的定义，对参数类型做检查，只允许整数和浮点数类型的参数。数据类型检查可以用内置函数isinstance()实现：

In [6]:
def my_abs(x):
    if not isinstance(x, (int, float)):
        raise TypeError('bad operand type')
    if x >= 0:
        return x
    else:
        return -x
my_abs('A')
abs('A')

TypeError: bad operand type

## 默认参数
定义默认参数要牢记一点：默认参数必须指向不变对象！str、None这样的不变对象

In [ ]:
def add_end(L=None):
    if L is None:
        L = []
    L.append('END')
    return L

在Python函数中，还可以定义可变参数。顾名思义，可变参数就是传入的参数个数是可变的，可以是1个、2个到任意个，还可以是0个。

我们以数学题为例子，给定一组数字a，b，c……，请计算a**2 + b**2 + c**2 + ……。

要定义出这个函数，我们必须确定输入的参数。由于参数个数不确定，我们首先想到可以把a，b，c……作为一个list或tuple传进来，这样，函数可以定义如下：

In [7]:
def calc(numbers):
    sum = 0
    for n in numbers:
        sum = sum + n * n
    return sum
calc((1, 3, 5, 7))

84

## * tuple 
把函数的参数改为可变参数：
定义可变参数和定义一个list或tuple参数相比，仅仅在参数前面加了一个 * 号。在函数内部，参数numbers接收到的是一个tuple，因此，函数代码完全不变。但是，调用该函数时，可以传入任意个参数，包括0个参数：

In [8]:
def calc(*numbers):
    sum = 0
    for n in numbers:
        sum = sum + n * n
    return sum
calc(1, 3, 5, 7)

84

Python允许你在list或tuple前面加一个* 号，把list或tuple的元素变成可变参数传进去：

In [9]:
calc(*[1, 2, 3])

14

可变参数允许你传入0个或任意个参数，这些可变参数在函数调用时自动组装为一个tuple。而关键字参数允许你传入0个或任意个含参数名的参数，这些关键字参数在函数内部自动组装为dict。

In [11]:
def person(name, age, **kw):
    print('name:', name, 'age:', age, 'other:', kw)
person('Michael', 30)
person('Bob', 35, city='Beijing')
person('Adam', 45, gender='M', job='Engineer')
extra = {'city': 'Beijing', 'job': 'Engineer'}
person('Jack', 24, **extra)

name: Michael age: 30 other: {}
name: Bob age: 35 other: {'city': 'Beijing'}
name: Adam age: 45 other: {'gender': 'M', 'job': 'Engineer'}
name: Jack age: 24 other: {'city': 'Beijing', 'job': 'Engineer'}


## 函数参数
在Python中定义函数，可以用必选参数、默认参数、可变参数、关键字参数和命名关键字参数，这5种参数都可以组合使用。但是请注意，参数定义的顺序必须是：必选参数、默认参数、可变参数、命名关键字参数和关键字参数。

In [ ]:
def f1(a, b, c=0, *args, **kw):
    print('a =', a, 'b =', b, 'c =', c, 'args =', args, 'kw =', kw)

def f2(a, b, c=0, *, d, **kw):
    print('a =', a, 'b =', b, 'c =', c, 'd =', d, 'kw =', kw)

f1(1, 2)
f1(1, 2, c=3)

*args是可变参数，args接收的是一个tuple；

**kw是关键字参数，kw接收的是一个dict。

以及调用函数时如何传入可变参数和关键字参数的语法：

可变参数既可以直接传入：func(1, 2, 3)，又可以先组装list或tuple，再通过*args传入：func(*(1, 2, 3))；

关键字参数既可以直接传入：func(a=1, b=2)，又可以先组装dict，再通过**kw传入：func(**{'a': 1, 'b': 2})。

## 函数递归
在函数内部，可以调用其他函数。如果一个函数在内部调用自身本身，这个函数就是递归函数。
举个例子，我们来计算阶乘n! = 1 x 2 x 3 x ... x n，用函数fact(n)表示，可以看出：
fact(n) = n! = 1 x 2 x 3 x ... x (n-1) x n = (n-1)! x n = fact(n-1) x n
所以，fact(n)可以表示为n x fact(n-1)，只有n=1时需要特殊处理。
于是，fact(n)用递归的方式写出来就是：

In [14]:
def fact(n):
    if n==1:
        return 1
    return n * fact(n - 1)
fact(5)

120

计算过程如下：
===> fact(5)
===> 5 * fact(4)
===> 5 * (4 * fact(3))
===> 5 * (4 * (3 * fact(2)))
===> 5 * (4 * (3 * (2 * fact(1))))
===> 5 * (4 * (3 * (2 * 1)))
===> 5 * (4 * (3 * 2))
===> 5 * (4 * 6)
===> 5 * 24
===> 120

使用递归函数需要注意防止栈溢出。在计算机中，函数调用是通过栈（stack）这种数据结构实现的，每当进入一个函数调用，栈就会加一层栈帧，每当函数返回，栈就会减一层栈帧。由于栈的大小不是无限的，所以，递归调用的次数过多，会导致栈溢出。可以试试fact(1000)：

## 尾递归
解决递归调用栈溢出的方法是通过尾递归优化，事实上尾递归和循环的效果是一样的，所以，把循环看成是一种特殊的尾递归函数也是可以的。

尾递归是指，在函数返回的时候，调用自身本身，并且，return语句不能包含表达式。这样，编译器或者解释器就可以把尾递归做优化，使递归本身无论调用多少次，都只占用一个栈帧，不会出现栈溢出的情况。

In [15]:
def fact(n):
    return fact_iter(n, 1)

def fact_iter(num, product):
    if num == 1:
        return product
    return fact_iter(num - 1, num * product)
fact(5)

120

===> fact_iter(5, 1)
===> fact_iter(4, 5)
===> fact_iter(3, 20)
===> fact_iter(2, 60)
===> fact_iter(1, 120)
===> 120

In [16]:
[x * x for x in range(1, 11) if x % 2 == 0] #列表生成式

[4, 16, 36, 64, 100]

 列出当前目录下的所有文件和目录名，可以通过一行代码实现：

In [ ]:
[d for d in os.listdir('.')] # os.listdir可以列出文件和目录

要创建一个generator，有很多种方法。第一种方法很简单，只要把一个列表生成式的[]改成()，就创建了一个generator：generator保存的是算法

In [ ]:
L = [x * x for x in range(10)]
L
g = (x * x for x in range(10))
g

斐波拉契数列用列表生成式写不出来，但是，用函数把它打印出来却很容易（除第一个和第二个数外，任意一个数都可由前两个数相加得到：）

In [18]:
def fib(max):
    n, a, b = 0, 0, 1
    while n < max:
        print(b)
        a, b = b, a + b
        n = n + 1
    return 'done'
fib(6)

1
1
2
3
5
8


'done'

a, b = b, a + b
相当于：
t = (b, a + b) # t是一个tuple
a = t[0]
b = t[1]

要把fib函数变成generator，只需要把print(b)改为yield b就可以了：

In [ ]:
def fib(max):
    n, a, b = 0, 0, 1
    while n < max:
        yield b
        a, b = b, a + b
        n = n + 1
    return 'done'

最难理解的就是generator和函数的执行流程不一样。函数是顺序执行，遇到return语句或者最后一行函数语句就返回。而变成generator的函数，在每次调用next()的时候执行，遇到yield语句返回，再次执行时从上次返回的yield语句处继续执行。在执行过程中，遇到yield就中断，下次又继续执行。

## 迭代器=可迭代对象+生成器
可以直接作用于for循环的数据类型有以下几种：
一类是集合数据类型，如list、tuple、dict、set、str等；
一类是generator，包括生成器和带yield的generator function。
这些可以直接作用于for循环的对象统称为可迭代对象：Iterable。
生成器都是Iterator对象，但list、dict、str虽然是Iterable，却不是Iterator。
把list、dict、str等Iterable变成Iterator可以使用iter()函数
Iterator的计算是惰性的，只有在需要返回下一个数据时它才会计算。
Iterator甚至可以表示一个无限大的数据流

In [26]:
from collections.abc import Iterable   #判断对象是否可迭代
from collections.abc import Iterator   #判断对象是否迭代器
print(isinstance([], Iterable))
print(isinstance([], Iterator))
print(isinstance(iter([]), Iterator))
print(isinstance(iter([]), Iterable))

True
False
True
True


函数式编程的一个特点就是，允许把函数本身作为参数传入另一个函数，还允许返回一个函数！

# 高阶函数
## Map() Rduce()函数
map()函数接收两个参数，一个是函数，一个是Iterable，map将传入的函数依次作用到序列的每个元素，并把结果作为新的Iterator返回。

In [28]:
def f(x):
    return x*x
r = map(f, [1, 2, 3, 4, 5, 6, 7, 8, 9])
r
list(r)

[1, 4, 9, 16, 25, 36, 49, 64, 81]

reduce把一个函数作用在一个序列[x1, x2, x3, ...]上，这个函数必须接收两个参数，reduce把结果继续和序列的下一个元素做累积计算，其效果就是：
reduce(f, [x1, x2, x3, x4]) = f(f(f(x1, x2), x3), x4)

In [ ]:
from functools import reduce
def fn(x, y):
    return x * 10 + y
def char2num(s):
    digits = {'0': 0, '1': 1, '2': 2, '3': 3, '4': 4, '5': 5, '6': 6, '7': 7, '8': 8, '9': 9}
    return digits[s]
reduce(fn, map(char2num, '13579'))

In [ ]:
from functools import reduce
DIGITS = {'0': 0, '1': 1, '2': 2, '3': 3, '4': 4, '5': 5, '6': 6, '7': 7, '8': 8, '9': 9}
def str2int(s):
    def fn(x, y):
        return x * 10 + y
    def char2num(s):
        return DIGITS[s]
    return reduce(fn, map(char2num, s))

In [ ]:
from functools import reduce
DIGITS = {'0': 0, '1': 1, '2': 2, '3': 3, '4': 4, '5': 5, '6': 6, '7': 7, '8': 8, '9': 9}
def char2num(s):
    return DIGITS[s]
def str2int(s):
    return reduce(lambda x, y: x * 10 + y, map(char2num, s))

## filter()
Python内建的filter()函数用于过滤序列。filter()也接收一个函数和一个序列。和map()不同的是，filter()把传入的函数依次作用于每个元素，然后根据返回值是True还是False决定保留还是丢弃该元素。
可见用filter()这个高阶函数，关键在于正确实现一个“筛选”函数。
注意到filter()函数返回的是一个Iterator，也就是一个惰性序列，所以要强迫filter()完成计算结果，需要用list()函数获得所有结果并返回list。

In [ ]:
def is_odd(n):
    return n % 2 == 1
list(filter(is_odd, [1, 2, 4, 5, 6, 9, 10, 15]))

def not_empty(s):
    return s and s.strip()
list(filter(not_empty, ['A', '', 'B', None, 'C', '  ']))

## 函数作为函数返回值
当我们调用lazy_sum()时，返回的并不是求和结果，而是求和函数：

In [30]:
def lazy_sum(*args):
    def sum():
        ax = 0
        for n in args:
            ax = ax + n
        return ax
    return sum
f = lazy_sum(1, 3, 5, 7, 9)
f
f()

25

在函数lazy_sum中又定义了函数sum，并且，内部函数sum可以引用外部函数lazy_sum的参数和局部变量，当lazy_sum返回函数sum时，相关参数和变量都保存在返回的函数中，这种称为“闭包（Closure）”的程序结构拥有极大的威力。
请再注意一点，当我们调用lazy_sum()时，每次调用都会返回一个新的函数，即使传入相同的参数：f1()和f2()的调用结果互不影响。

In [ ]:
f1 = lazy_sum(1, 3, 5, 7, 9)
f2 = lazy_sum(1, 3, 5, 7, 9)
f1==f2

返回的函数并没有立刻执行，而是直到调用了f()才执行。

In [33]:
def count():
    fs = []
    for i in range(1, 4):
        def f():
             return i*i
        fs.append(f)
    return fs
f1, f2, f3 = count() #返回闭包时牢记一点：返回函数不要引用任何循环变量，或者后续会发生变化的变量。
f1()
f2()
f3()

9

In [32]:
def count():
    def f(j):
        def g():
            return j*j
        return g
    fs = []
    for i in range(1, 4):
        fs.append(f(i)) # f(i)立刻被执行，因此i的当前值被传入f()
    return fs
f1, f2, f3 = count() 
f1()
f2()
f3()

25

## 匿名函数

In [34]:
list(map(lambda x: x * x, [1, 2, 3, 4, 5, 6, 7, 8, 9]))

[1, 4, 9, 16, 25, 36, 49, 64, 81]

关键字lambda表示匿名函数，冒号前面的x表示函数参数。
匿名函数有个限制，就是只能有一个表达式，不用写return，返回值就是该表达式的结果。
用匿名函数有个好处，因为函数没有名字，不必担心函数名冲突。此外，匿名函数也是一个函数对象，也可以把匿名函数赋值给一个变量，再利用变量来调用该函数：

In [36]:
f = lambda x: x * x
f
f(5)


25

In [38]:
def build(x, y):
    return lambda: x * x + y * y
b=build(2,3)
b
b()

13

由于函数也是一个对象，而且函数对象可以被赋值给变量，所以，通过变量也能调用该函数。

In [ ]:
def now():
    print('2015-3-25')
f = now
f()

函数对象有一个__name__属性，可以拿到函数的名字：

In [ ]:
now.__name__
f.__name__

现在，假设我们要增强now()函数的功能，比如，在函数调用前后自动打印日志，但又不希望修改now()函数的定义，这种在代码运行期间动态增加功能的方式，称之为“装饰器”（Decorator）。
本质上，decorator就是一个返回函数的高阶函数。所以，我们要定义一个能打印日志的decorator，可以定义如下：

In [39]:
def log(func):
    def wrapper(*args, **kw):
        print('call %s():' % func.__name__)
        return func(*args, **kw)
    return wrapper
#观察上面的log，因为它是一个decorator，所以接受一个函数作为参数，并返回一个函数。我们要借助Python的@语法，把decorator置于函数的定义处：
@log
def now():
    print('2015-3-25')
now() #调用now()函数，不仅会运行now()函数本身，还会在运行now()函数前打印一行日志：

call now():
2015-3-25


由于log()是一个decorator，返回一个函数，所以，原来的now()函数仍然存在，只是现在同名的now变量指向了新的函数，于是调用now()将执行新函数，即在log()函数中返回的wrapper()函数。
wrapper()函数的参数定义是(* args, ** kw)，因此，wrapper()函数可以接受任意参数的调用。在wrapper()函数内，首先打印日志，再紧接着调用原始函数。

## 偏函数
Python的functools模块提供了很多有用的功能，其中一个就是偏函数（Partial function）。要注意，这里的偏函数和数学意义上的偏函数不一样。
简单总结functools.partial的作用就是，把一个函数的某些参数给固定住（也就是设置默认值），返回一个新的函数，调用这个新函数会更简单。

In [ ]:
def int2(x, base=2):
    return int(x, base)
#相当于
import functools
int2 = functools.partial(int, base=2)
int2('1000000')

# 模块
一个.py文件就称之为一个模块（Module）
按目录来组织模块的方法，称为包（Package）
一个abc.py的文件就是一个名字叫abc的模块，一个xyz.py的文件就是一个名字叫xyz的模块。
现在，假设我们的abc和xyz这两个模块名字与其他模块冲突了，于是我们可以通过包来组织模块，避免冲突。方法是选择一个顶层包名，比如mycompany，按照如下目录存放：
mycompany
├─ __init__.py
├─ abc.py
└─ xyz.py
请注意，每一个包目录下面都会有一个__init__.py的文件，这个文件是必须存在的，否则，Python就把这个目录当成普通目录，而不是一个包。__init__.py可以是空文件，也可以有Python代码，因为__init__.py本身就是一个模块，而它的模块名就是mycompany。
类似的，可以有多级目录，组成多级层次的包结构。比如如下的目录结构：
mycompany
 ├─ web
 │  ├─ __init__.py
 │  ├─ utils.py
 │  └─ www.py
 ├─ __init__.py
 ├─ abc.py
 └─ xyz.py
创建自己的模块时，要注意：
模块名要遵循Python变量命名规范，不要使用中文、特殊字符；
模块名不要和系统模块名冲突，最好先查看系统是否已存在该模块，检查方法是在Python交互环境执行import abc，若成功则说明系统存在此模块。

In [8]:
#!/usr/bin/env python3  第1行注释可以让这个hello.py文件直接在Unix/Linux/Mac上运行
# -*- coding: utf-8 -*-        第2行注释表示.py文件本身使用标准UTF-8编码；

' a test module ' #任何模块代码的第一个字符串都被视为模块的文档注释

__author__ = 'Michael Liao'  #__author__变量把作者写进去，这样当你公开源代码后别人就可以瞻仰你的大名；
#导入sys模块后，我们就有了变量sys指向该模块，利用sys这个变量，就可以访问sys模块的所有功能。
import sys
#
def test():
    args = sys.argv
    print(args)
    if len(args)==1:
        print('Hello, world!')
    elif len(args)==2:
        print('Hello, %s!' % args[1])
    else:
        print('Too many arguments!')
         
if __name__=='__main__':
    test()


['D:\\Aconda\\lib\\site-packages\\ipykernel_launcher.py', '-f', 'C:\\Users\\Administrator\\AppData\\Roaming\\jupyter\\runtime\\kernel-743f1fc9-1038-43cd-a221-b513aa410dda.json']
Too many arguments!


当我们在命令行运行hello模块文件时，Python解释器把一个特殊变量__name__置为__main__，而如果在其他地方导入该hello模块时，if判断将失败，因此，这种if测试可以让一个模块通过命令行运行时执行一些额外的代码，最常见的就是运行测试。

## 作用域
在一个模块中，我们可能会定义很多函数和变量，但有的函数和变量我们希望给别人使用，有的函数和变量我们希望仅仅在模块内部使用。在Python中，是通过_前缀来实现的。
类似_xxx和__xxx这样的函数或变量就是非公开的（private），不应该被直接引用
private函数和变量“不应该”被直接引用，而不是“不能”被直接引用，是因为Python并没有一种方法可以完全限制访问private函数或变量，但是，从编程习惯上不应该引用private函数或变量。

In [11]:
def _private_1(name):
    return 'Hello, %s' % name

def _private_2(name):
    return 'Hi, %s' % name

def greeting(name):
    if len(name) > 3:
        return _private_1(name)
    else:
        return _private_2(name)
greeting('aaaaaa')
# 在模块里公开greeting()函数，而把内部逻辑用private函数隐藏起来了
# 外部不需要引用的函数全部定义成private，只有外部需要引用的函数才定义为public。

'Hello, aaaaaa'

## 安装第三方库
在Python中，安装第三方模块，是通过包管理工具pip完成的。
如果你正在使用Mac或Linux，安装pip本身这个步骤就可以跳过了。
如果你正在使用Windows，请参考安装Python一节的内容，确保安装时勾选了pip和Add python.exe to Path。
在命令提示符窗口下尝试运行pip，如果Windows提示未找到命令，可以重新运行安装程序添加pip。
用pip一个一个安装费时费力，还需要考虑兼容性。我们推荐直接使用Anaconda，这是一个基于Python的数据处理和科学计算平台，它已经内置了许多非常有用的第三方库，我们装上Anaconda，Anaconda会把系统Path中的python指向自己自带的Python，并且，Anaconda安装的第三方模块会安装在Anaconda自己的路径下，不影响系统已安装的Python目录。
当我们试图加载一个模块时，Python会在指定的路径下搜索对应的.py文件，如果找不到，就会报错
默认情况下，Python解释器会搜索当前目录、所有已安装的内置模块和第三方模块，搜索路径存放在sys模块的path变量中：

In [12]:
import sys
sys.path

['',
 'C:\\Users\\Administrator',
 'D:\\Aconda\\python37.zip',
 'D:\\Aconda\\DLLs',
 'D:\\Aconda\\lib',
 'D:\\Aconda',
 'D:\\Aconda\\lib\\site-packages',
 'D:\\Aconda\\lib\\site-packages\\win32',
 'D:\\Aconda\\lib\\site-packages\\win32\\lib',
 'D:\\Aconda\\lib\\site-packages\\Pythonwin',
 'D:\\Aconda\\lib\\site-packages\\IPython\\extensions',
 'C:\\Users\\Administrator\\.ipython']

如果我们要添加自己的搜索目录，有两种方法：
一是直接修改sys.path，添加要搜索的目录：

In [13]:
import sys
sys.path.append('/Users/michael/my_py_scripts')
sys.path

['',
 'C:\\Users\\Administrator',
 'D:\\Aconda\\python37.zip',
 'D:\\Aconda\\DLLs',
 'D:\\Aconda\\lib',
 'D:\\Aconda',
 'D:\\Aconda\\lib\\site-packages',
 'D:\\Aconda\\lib\\site-packages\\win32',
 'D:\\Aconda\\lib\\site-packages\\win32\\lib',
 'D:\\Aconda\\lib\\site-packages\\Pythonwin',
 'D:\\Aconda\\lib\\site-packages\\IPython\\extensions',
 'C:\\Users\\Administrator\\.ipython',
 '/Users/michael/my_py_scripts']

第二种方法是设置环境变量PYTHONPATH，该环境变量的内容会被自动添加到模块搜索路径中。设置方式与设置Path环境变量类似。注意只需要添加你自己的搜索路径，Python自己本身的搜索路径不受影响。

# 面向对象编程
## 类和实例
类是抽象的模板，比如Student类，而实例是根据类创建出来的一个个具体的“对象”，每个对象都拥有相同的方法，但各自的数据可能不同。

In [14]:
class Student(object):
    pass

class后面紧接着是类名，即Student，类名通常是大写开头的单词，紧接着是(object)，表示该类是从哪个类继承下来的，继承的概念我们后面再讲，通常，如果没有合适的继承类，就使用object类，这是所有类最终都会继承的类
定义好了Student类，就可以根据Student类创建出Student的实例，创建实例是通过类名+()实现的：
变量bart指向的就是一个Student的实例，后面的0x10a67a590是内存地址，每个object的地址都不一样，而Student本身则是一个类。

In [18]:
bart = Student()
print(bart)
print(Student)
bart.name = 'Bart Simpson'  #自由地给一个实例变量绑定属性
bart.name

<class '__main__.Student'>


'Bart Simpson'

可以在创建实例的时候，把一些我们认为必须绑定的属性强制填写进去。通过定义一个特殊的__init__方法，在创建实例的时候，就把name，score等属性绑上去：

In [20]:
class Student(object):

    def __init__(self, name, score):
        self.name = name
        self.score = score
bart = Student('Bart Simpson', 59)
bart.name

'Bart Simpson'

注意到__init__方法的第一个参数永远是self，表示创建的实例本身，因此，在__init__方法内部，就可以把各种属性绑定到self，因为self就指向创建的实例本身。
有了__init__方法，在创建实例的时候，就不能传入空的参数了，必须传入与__init__方法匹配的参数，但self不需要传，Python解释器自己会把实例变量传进去
和普通的函数相比，在类中定义的函数只有一点不同，就是第一个参数永远是实例变量self，并且，调用时，不用传递该参数。除此之外，类的方法和普通函数没有什么区别，所以，你仍然可以用默认参数、可变参数、关键字参数和命名关键字参数。

### 数据封装
面向对象编程的一个重要特点就是数据封装。实例本身就拥有这些数据，要访问这些数据，就没有必要从外面的函数去访问，可以直接在Student类的内部定义访问数据的函数，这样，就把“数据”给封装起来了。这些封装数据的函数是和Student类本身是关联起来的，我们称之为类的方法

In [23]:
class Student(object):
    def __init__(self, name, score):
        self.name = name
        self.score = score
    def print_score(self):
        print('%s: %s' % (self.name, self.score))
    def get_grade(self):
        if self.score >= 90:
            return 'A'
        elif self.score >= 60:
            return 'B'
        else:
            return 'C'
bart=Student('Bart Simpson',9)
bart.print_score()
bart.get_grade()

Bart Simpson: 9


'C'

和静态语言不同，Python允许对实例变量绑定任何数据，也就是说，对于两个实例变量，虽然它们都是同一个类的不同实例，但拥有的变量名称都可能不同：
外部代码还是可以自由地修改一个实例的name、score属性,甚至添加属性age

In [24]:
bart = Student('Bart Simpson', 59)
lisa = Student('Lisa Simpson', 87)
bart.age = 8
bart.age
lisa.age

AttributeError: 'Student' object has no attribute 'age'

如果要让内部属性不被外部访问，可以把属性的名称前加上两个下划线__，在Python中，实例的变量名如果以__开头，就变成了一个私有变量（private），只有内部可以访问，外部不能访问

In [27]:
class Student(object):
    def __init__(self, name, score):
        self.__name = name
        self.__score = score
    def print_score(self):
        print('%s: %s' % (self.__name, self.__score))    
    def get_name(self):             #外部代码要获取name和score
        return self.__name
    def get_score(self):
        return self.__score
    def set_score(self, score):     #允许外部代码修改score
        if 0 <= score <= 100:
            self.__score = score
        else:
            raise ValueError('bad score')  #在方法中，可以对参数做检查，避免传入无效的参数：
#改完后，对于外部代码来说，没什么变动，但是已经无法从外部访问实例变量.__name和实例变量.__score了：
bart = Student('Bart Simpson', 59)
#bart.__name
bart.set_score(60)
bart.get_score()

60

1)在Python中，变量名类似__xxx__的，也就是以双下划线开头，并且以双下划线结尾的，是特殊变量，特殊变量是可以直接访问的，不是private变量，所以，不能用__name__、__score__这样的变量名。
2)有些时候，你会看到以一个下划线开头的实例变量名，比如_name，这样的实例变量外部是可以访问的，但是，按照约定俗成的规定，当你看到这样的变量时，意思就是，“虽然我可以被访问，但是，请把我视为私有变量，不要随意访问”。
3）双下划线开头的实例变量是不是一定不能从外部访问呢？其实也不是。不能直接访问__name是因为Python解释器对外把__name变量改成了_Student__name，所以，仍然可以通过_Student__name来访问__name变量：（不同版本的Python解释器可能会把__name改成不同的变量名）

In [28]:
bart._Student__name

'Bart Simpson'

错误写法：表面上看，外部代码“成功”地设置了__name变量，但实际上这个__name变量和class内部的__name变量不是一个变量！内部的__name变量已经被Python解释器自动改成了_Student__name，而外部代码给bart新增了一个__name变量。

In [31]:
bart = Student('Bart Simpson', 59)
bart.get_name()
bart.__name = 'New Name'
print(bart.__name)
print(bart.get_name())

New Name
Bart Simpson


## 继承和多态
在OOP程序设计中，当我们定义一个class的时候，可以从某个现有的class继承，新的class称为子类（Subclass），而被继承的class称为基类、父类或超类（Base class、Super class）。子类获得了父类的全部功能。

In [32]:
class Animal(object):
    def run(self):
        print('Animal is running...')
class Dog(Animal):
    def run(self):
        print('Dog is running...')
    def eat(self):
        print('Eating meat...')
class Cat(Animal):
    def run(self):
        print('Cat is running...')
dog = Dog()
dog.run()
#当子类和父类都存在相同的run()方法时，我们说，子类的run()覆盖了父类的run()
cat = Cat()
cat.run()

Dog is running...
Cat is running...


### 多态
当我们定义一个class的时候，我们实际上就定义了一种“数据类型”。我们定义的数据类型和Python自带的数据类型，比如str、list、dict没什么两样：

In [ ]:
a = list() # a是list类型
b = Animal() # b是Animal类型
c = Dog() # c是Dog类型
#判断一个变量是否是某个类型可以用isinstance()判断：
isinstance(a, list)
isinstance(c, Animal)
isinstance(c, Dog)
#在继承关系中，如果一个实例的数据类型是某个子类，那它的数据类型也可以被看做是父类。但是，反过来就不行
isinstance(b, Dog)

In [35]:
def run_twice(a): #编写一个函数，这个函数接受一个Animal类型的变量
    a.run()
    a.run()
run_twice(Animal())    #当我们传入Animal的实例时，run_twice()就打印出
run_twice(Cat())       #当我们传入Cat的实例时，run_twice()就打印出：

class Tortoise(Animal):#再定义一个Tortoise类型，也从Animal派生：
    def run(self):
        print('Tortoise is running slowly...')
run_twice(Tortoise())

Animal is running...
Animal is running...
Cat is running...
Cat is running...
Tortoise is running slowly...
Tortoise is running slowly...


新增一个Animal的子类，不必对run_twice()做任何修改，实际上，任何依赖Animal作为参数的函数或者方法都可以不加修改地正常运行，原因就在于多态。
多态的好处就是，当我们需要传入Dog、Cat、Tortoise……时，我们只需要接收Animal类型就可以了，因为Dog、Cat、Tortoise……都是Animal类型，然后，按照Animal类型进行操作即可。由于Animal类型有run()方法，因此，传入的任意类型，只要是Animal类或者子类，就会自动调用实际类型的run()方法，这就是多态的意思：
对于一个变量，我们只需要知道它是Animal类型，无需确切地知道它的子类型，就可以放心地调用run()方法，而具体调用的run()方法是作用在Animal、Dog、Cat还是Tortoise对象上，由运行时该对象的确切类型决定，这就是多态真正的威力：调用方只管调用，不管细节，而当我们新增一种Animal的子类时，只要确保run()方法编写正确，不用管原来的代码是如何调用的。这就是著名的“开闭”原则：
1）对扩展开放：允许新增Animal子类；
2）对修改封闭：不需要修改依赖Animal类型的run_twice()等函数。
继承还可以一级一级地继承下来，就好比从爷爷到爸爸、再到儿子这样的关系。而任何类，最终都可以追溯到根类object，这些继承关系看上去就像一颗倒着的树。比如如下的继承树：
                ┌───────────────┐
                │    object     │
                └───────────────┘
                        │
           ┌────────────┴────────────┐
           │                         │
           ▼                         ▼
    ┌─────────────┐           ┌─────────────┐
    │   Animal    │           │    Plant    │
    └─────────────┘           └─────────────┘
           │                         │
     ┌─────┴──────┐            ┌─────┴──────┐
     │            │            │            │
     ▼            ▼            ▼            ▼
┌─────────┐  ┌─────────┐  ┌─────────┐  ┌─────────┐
│   Dog   │  │   Cat   │  │  Tree   │  │ Flower  │
└─────────┘  └─────────┘  └─────────┘  └─────────┘

### 静态语言 vs 动态语言
对于静态语言（例如Java）来说，如果需要传入Animal类型，则传入的对象必须是Animal类型或者它的子类，否则，将无法调用run()方法。
对于Python这样的动态语言来说，则不一定需要传入Animal类型。我们只需要保证传入的对象有一个run()方法就可以了：
动态语言的“鸭子类型”，它并不要求严格的继承体系，一个对象只要“看起来像鸭子，走起路来像鸭子”，那它就可以被看做是鸭子。

In [34]:
class Timer(object):
    def run(self):
        print('Start...')
run_twice(Timer())

Start...
Start...


## 获取对象信息
### 使用type()
基本类型都可以用type()判断：

In [36]:
type(123)
type('str')
type(None)


NoneType

In [ ]:
type(abs)
type(a)

In [ ]:
#如果要判断一个对象是否是函数
import types
def fn():
    pass
type(fn)==types.FunctionType              #函数
type(abs)==types.BuiltinFunctionType      #内置函数
type(lambda x: x)==types.LambdaType       #Lambda函数
type((x for x in range(10)))==types.GeneratorType   #生成器

### 使用isinstance
对于class的继承关系来说，使用type()就很不方便。要判断class的类型，可以使用isinstance()函数。
如继承关系为：object -> Animal -> Dog -> Husky

In [38]:
class Husky(Dog):
    pass
a = Animal()
d = Dog()
h = Husky()
isinstance(h, Animal)
isinstance(d, Husky)

isinstance([1, 2, 3], (list, tuple))  #判断一个变量是否是某些类型中的一种

### 使用dir()
要获得一个对象的所有属性和方法，可以使用dir()函数，它返回一个包含字符串的list

In [39]:
dir('ABC')

['__add__',
 '__class__',
 '__contains__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getitem__',
 '__getnewargs__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__mod__',
 '__mul__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__rmod__',
 '__rmul__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 'capitalize',
 'casefold',
 'center',
 'count',
 'encode',
 'endswith',
 'expandtabs',
 'find',
 'format',
 'format_map',
 'index',
 'isalnum',
 'isalpha',
 'isascii',
 'isdecimal',
 'isdigit',
 'isidentifier',
 'islower',
 'isnumeric',
 'isprintable',
 'isspace',
 'istitle',
 'isupper',
 'join',
 'ljust',
 'lower',
 'lstrip',
 'maketrans',
 'partition',
 'replace',
 'rfind',
 'rindex',
 'rjust',
 'rpartition',
 'rsplit',
 'rstrip',
 'split',
 'splitlines',
 'startswith',
 'strip',
 'swapcase',
 'title',
 'translate',
 'upper',


类似__xxx__的属性和方法在Python中都是有特殊用途的，比如__len__方法返回长度。在Python中，如果你调用len()函数试图获取一个对象的长度，实际上，在len()函数内部，它自动去调用该对象的__len__()方法，所以，下面的代码是等价的：
剩下的都是普通属性或方法

In [40]:
len('ABC')
#等价于
'ABC'.__len__()

3

自己写的类，如果也想用len(myObj)的话，就自己写一个__len__()方法：

In [44]:
class MyDog(object):
    def __len__(self):
        return 10
dog = MyDog()
print(len(dog))
print(len('ssss'))

10
4


仅仅把属性和方法列出来是不够的，配合getattr()、setattr()以及hasattr()，我们可以直接操作一个对象的状态：

In [46]:
class MyObject(object):
    def __init__(self):
        self.x=9
    def power(self):
        return self.x * self.x
obj= MyObject()
obj

In [50]:
hasattr(obj, 'x') #有属性'x'吗？
setattr(obj, 'y', 19) # 设置一个属性'y'
getattr(obj, 'y') # 获取属性'y'
#getattr(obj, 'z') #试图获取不存在的属性，会抛出AttributeError的错误：
#可以传入一个default参数，如果属性不存在，就返回默认值:
getattr(obj, 'z', 404) # 获取属性'z'，如果不存在，返回默认值404

404

getattr()以及hasattr()可以作用于对象的方法

In [ ]:
hasattr(obj, 'power') # 有属性'power'吗？
fn=getattr(obj, 'power') # 获取属性'power'
fn() # 调用fn()与调用obj.power()是一样的

通过内置的一系列函数，我们可以对任意一个Python对象进行剖析，拿到其内部的数据。要注意的是，只有在不知道对象信息的时候，我们才会去获取对象信息。如果可以直接写：
假设我们希望从文件流fp中读取图像，我们首先要判断该fp对象是否存在read方法，如果存在，则该对象是一个流，如果不存在，则无法读取。hasattr()就派上了用场。
请注意，在Python这类动态语言中，根据鸭子类型，有read()方法，不代表该fp对象就是一个文件流，它也可能是网络流，也可能是内存中的一个字节流，但只要read()方法返回的是有效的图像数据，就不影响读取图像的功能。

In [ ]:
def readImage(fp):
    if hasattr(fp, 'read'):
        return readData(fp)
    return None

## 实例属性
由于Python是动态语言，根据类创建的实例可以任意绑定属性。
给实例绑定属性的方法是通过实例变量，或者通过self变量：

In [ ]:
class Student(object):
    def __init__(self, name):
        self.name = name
s = Student('Bob')
s.score = 90

Student类本身需要绑定一个属性

在编写程序的时候，千万不要对实例属性和类属性使用相同的名字，因为相同名称的实例属性将屏蔽掉类属性，但是当你删除实例属性后，再使用相同的名称，访问到的将是类属性。

In [51]:
class Student(object):
    name = 'Student'   #类属性，这个属性虽然归类所有，但类的所有实例都可以访问到
s = Student() # 创建实例s
s.name

'Student'

In [52]:
class Student(object):
    name = 'Student'
    def __init__(self, name):
        self.name = name
s = Student('Bob')
s.name

'Bob'

实例属性属于各个实例所有，互不干扰；
类属性属于类所有，所有实例共享一个属性；
不要对实例属性和类属性使用相同的名字，否则将产生难以发现的错误。

# 面向对象高级编程
给实例绑定一个方法： 

In [54]:
class Student(object):
    pass
s = Student()
def set_age(self, age): # 定义一个函数作为实例方法
    self.age = age
from types import MethodType
s.set_age = MethodType(set_age, s) # 给实例绑定一个方法
s.set_age(25) # 调用实例方法
s.age # 测试结果
#但是，给一个实例绑定的方法，对另一个实例是不起作用的：
#s2 = Student() # 创建新的实例
#s2.set_age(25) # 尝试调用方法

25

In [57]:
#为了给所有实例都绑定方法，可以给class绑定方法：
def set_score(self, score):
    self.score = score
Student.set_score = set_score #给class绑定方法后，所有实例均可调用：
s2=Student()
s2.set_score(99)
s2.score
#通常情况下，上面的set_score方法可以直接定义在class中，但动态绑定允许我们在程序运行的过程中动态给class加上功能，
#这在静态语言中很难实现。

99

### 使用__slots__
限制实例的属性,Python允许在定义class的时候，定义一个特殊的__slots__变量，来限制该class实例能添加的属性：

In [58]:
class Student(object):
    __slots__ = ('name', 'age') # 用tuple定义允许绑定的属性名称
s = Student() # 创建新的实例
s.name = 'Michael' # 绑定属性'name'
s.score = 99 # 绑定属性'score'

AttributeError: 'Student' object has no attribute 'score'

使用__slots__要注意，__slots__定义的属性仅对当前类实例起作用，对继承的子类是不起作用

In [65]:
class GraduateStudent(Student):
    __slots__ =()
    pass
g = GraduateStudent()
g.score = 9999
g.score
#除非在子类中也定义__slots__，这样，子类实例允许定义的属性就是自身的__slots__加上父类的__slots__。

9999

### 使用@property
在绑定属性时，如果我们直接把属性暴露出去，虽然写起来很简单，但是，没办法检查参数，导致可以把成绩随便改；为了限制score的范围，可以通过一个set_score()方法来设置成绩，再通过一个get_score()来获取成绩，这样，在set_score()方法里，就可以检查参数：

In [ ]:
class Student(object):

    def get_score(self):
         return self._score

    def set_score(self, value):
        if not isinstance(value, int):
            raise ValueError('score must be an integer!')
        if value < 0 or value > 100:
            raise ValueError('score must between 0 ~ 100!')
        self._score = value
s = Student()
s.set_score(60) # ok!
s.get_score()

上面的调用方法又略显复杂，没有直接用属性这么直接简单。
有没有既能检查参数，又可以用类似属性这样简单的方式来访问类的变量呢？

装饰器（decorator）可以给函数动态加上功能，对于类的方法，装饰器一样起作用。Python内置的@property装饰器就是负责把一个方法变成属性调用的：

In [ ]:
class Student(object):

    @property                  #把一个getter方法变成属性,用于获取属性
    def score(self):
        return self._score
                                #@property本身又创建了另一个装饰器@score.setter，
    @score.setter               #负责把一个setter方法变成属性赋值                 
    def score(self, value):
        if not isinstance(value, int):
            raise ValueError('score must be an integer!')
        if value < 0 or value > 100:
            raise ValueError('score must between 0 ~ 100!')
        self._score = value
#可控的属性操作
s = Student()
s.score = 60
s.score 

定义只读属性，只定义getter方法，不定义setter方法就是一个只读属性：

In [ ]:
class Student(object):

    @property
    def birth(self):
        return self._birth

    @birth.setter
    def birth(self, value):
        self._birth = value

    @property
    def age(self):
        return 2015 - self._birth
#上面的birth是可读写属性，而age就是一个只读属性，因为age可以根据birth和当前时间计算出来。
#@property广泛应用在类的定义中，可以让调用者写出简短的代码，同时保证对参数进行必要的检查，这样，程序运行时就减少了出错的可能性。

## 多重继承
继承是面向对象编程的一个重要的方式，因为通过继承，子类就可以扩展父类的功能
类的层次就复杂了：
                ┌───────────────┐
                │    Animal     │
                └───────────────┘
                        │
           ┌────────────┴────────────┐
           │                         │
           ▼                         ▼
    ┌─────────────┐           ┌─────────────┐
    │   Mammal    │           │    Bird     │
    └─────────────┘           └─────────────┘
           │                         │
     ┌─────┴──────┐            ┌─────┴──────┐
     │            │            │            │
     ▼            ▼            ▼            ▼
┌─────────┐  ┌─────────┐  ┌─────────┐  ┌─────────┐
│  MRun   │  │  MFly   │  │  BRun   │  │  BFly   │
└─────────┘  └─────────┘  └─────────┘  └─────────┘
     │            │            │            │
     │            │            │            │
     ▼            ▼            ▼            ▼
┌─────────┐  ┌─────────┐  ┌─────────┐  ┌─────────┐
│   Dog   │  │   Bat   │  │ Ostrich │  │ Parrot  │
└─────────┘  └─────────┘  └─────────┘  └─────────┘
如果要再增加“宠物类”和“非宠物类”，这么搞下去，类的数量会呈指数增长，很明显这样设计是不行的。
正确的做法是采用多重继承。首先，主要的类层次仍按照哺乳类和鸟类设计：

In [ ]:
class Animal(object):
    pass

# 大类:
class Mammal(Animal):
    pass

class Bird(Animal):
    pass

# 各种动物:
class Dog(Mammal):
    pass

class Bat(Mammal):
    pass

class Parrot(Bird):
    pass

class Ostrich(Bird):
    pass
#现在，我们要给动物再加上Runnable和Flyable的功能，只需要先定义好Runnable和Flyable的类：
class Runnable(object):
    def run(self):
        print('Running...')

class Flyable(object):
    def fly(self):
        print('Flying...')
#对于需要Runnable功能的动物，就多继承一个Runnable，例如Dog：
class Dog(Mammal, Runnable):
    pass
#对于需要Flyable功能的动物，就多继承一个Flyable，例如Bat：
class Bat(Mammal, Flyable):
    pass
#通过多重继承，一个子类就可以同时获得多个父类的所有功能。
#拓扑排序---->指导多重继承的查询顺序

### MixIn
在设计类的继承关系时，通常，主线都是单一继承下来的，例如，Ostrich继承自Bird。但是，如果需要“混入”额外的功能，通过多重继承就可以实现，比如，让Ostrich除了继承自Bird外，再同时继承Runnable。这种设计通常称之为MixIn。
为了更好地看出继承关系，我们把Runnable和Flyable改为RunnableMixIn和FlyableMixIn。类似的，你还可以定义出肉食动物CarnivorousMixIn和植食动物HerbivoresMixIn，让某个动物同时拥有好几个MixIn：
class Dog(Mammal, RunnableMixIn, CarnivorousMixIn):
    pass
    
比如，编写一个多进程模式的TCP服务，定义如下：
class MyTCPServer(TCPServer, ForkingMixIn):
    pass
这样一来，我们不需要复杂而庞大的继承链，只要选择组合不同的类的功能，就可以快速构造出所需的子类。
由于Python允许使用多重继承，因此，MixIn就是一种常见的设计。
只允许单一继承的语言（如Java）不能使用MixIn的设计。

## 定制类
看到类似__slots__这种形如__xxx__的变量或者函数名就要注意，这些在Python中是有特殊用途的。
__slots__我们已经知道怎么用了，__len__()方法我们也知道是为了能让class作用于len()函数。
除此之外，Python的class中还有许多这样有特殊用途的函数，可以帮助我们定制类。

In [2]:
### __str__
class Student(object):
    def __init__(self, name):
        self.name = name
print(Student('Michael'))
print(Student('bob').name)
#前者打印出一堆<__main__.Student object at 0x109afb190>，不好看。

bob


In [3]:
class Student(object):
    def __init__(self, name):
        self.name = name
    def __str__(self): #定义好__str__()方法，返回一个好看的字符串
        return 'Student object (name: %s)' % self.name
print(Student('Michael'))    
s = Student('Michael') #直接敲变量不用print，打印出来的实例还是不好看：


Student object (name: Michael)


直接显示变量调用的不是__str__()，而是__repr__()
两者的区别是__str__()返回用户看到的字符串，而__repr__()返回程序开发者看到的字符串，也就是说，__repr__()是为调试服务的。
解决办法是再定义一个__repr__()。但是通常__str__()和__repr__()代码都是一样的，所以，有个偷懒的写法：

In [4]:
class Student(object):
    def __init__(self, name):
        self.name = name
    def __str__(self):
        return 'Student object (name=%s)' % self.name
    __repr__ = __str__
s = Student('Michael')
s

Student object (name=Michael)

### __iter__
如果一个类想被用于for ... in循环，类似list或tuple那样，就必须实现一个__iter__()方法，该方法返回一个迭代对象，然后，Python的for循环就会不断调用该迭代对象的__next__()方法拿到循环的下一个值，直到遇到StopIteration错误时退出循环。
们以斐波那契数列为例，写一个Fib类，可以作用于for循环：

In [6]:
class Fib(object):
    def __init__(self):
        self.a, self.b = 0, 1 # 初始化两个计数器a，b

    def __iter__(self):
        return self # 实例本身就是迭代对象，故返回自己

    def __next__(self):  #or循环就会不断调用该迭代对象的__next__()方法拿到循环的下一个值
        self.a, self.b = self.b, self.a + self.b # 计算下一个值
        if self.a > 30: # 退出循环的条件
            raise StopIteration()    #到遇到StopIteration错误时退出循环
        return self.a # 返回下一个值
for n in Fib():
    print(n)

1
1
2
3
5
8
13
21


### __getitem__
Fib实例虽然能作用于for循环，看起来和list有点像，但是，把它当成list来使用还是不行，比如，取第5个元素：

In [ ]:
Fib()[5]

要表现得像list那样按照下标取出元素，需要实现__getitem__()方法：

In [11]:
class Fib(object):
    def __getitem__(self, n):
        a, b = 1, 1
        for x in range(n):
            a, b = b, a + b
        return a
f=Fib()
f[5]
#切片方法对于Fib却报错。原因是__getitem__()传入的参数可能是一个int，
#也可能是一个切片对象slice，所以要做判断：

8

In [ ]:
class Fib(object):
    def __getitem__(self, n):
        if isinstance(n, int): # n是索引
            a, b = 1, 1
            for x in range(n):
                a, b = b, a + b
            return a           #返回值
        if isinstance(n, slice): # n是切片
            start = n.start
            stop = n.stop
            if start is None:
                start = 0
            a, b = 1, 1
            L = []
            for x in range(stop):
                if x >= start:
                    L.append(a)
                a, b = b, a + b
            return L        #返回列表
#但是没有对step参数作处理,也没有对负数作处理
#所以，要正确实现一个__getitem__()还是有很多工作要做的。

此外，如果把对象看成dict，__getitem__()的参数也可能是一个可以作key的object，例如str。
与之对应的是__setitem__()方法，把对象视作list或dict来对集合赋值。最后，还有一个__delitem__()方法，用于删除某个元素。
总之，通过上面的方法，我们自己定义的类表现得和Python自带的list、tuple、dict没什么区别，这完全归功于动态语言的“鸭子类型”，不需要强制继承某个接口。

### __getattr__
正常情况下，当我们调用类的方法或属性时，如果不存在，就会报错。比如定义Student类：

In [ ]:
class Student(object):
    def __init__(self):
        self.name = 'Michael'
#调用name属性，没问题，但是，调用不存在的score属性，就有问题了：
s = Student()
print(s.name)
print(s.score)

写一个__getattr__()方法，动态返回一个属性。

In [ ]:
class Student(object):
    def __init__(self):
        self.name = 'Michael'
    def __getattr__(self, attr):
        if attr=='score':
            return 99
#当调用不存在的属性时，比如score，Python解释器会试图调用__getattr__(self, 'score')来尝试获得属性，
#这样，我们就有机会返回score的值：
s = Student()
s.score

In [15]:
#返回函数也是完全可以的：
class Student(object):
    def __getattr__(self, attr):
        if attr=='age':
            return lambda: 25
s=Student()
s.age()    #注意调用方式
print(s.socre)

None


注意，只有在没有找到属性的情况下，才调用__getattr__，已有的属性，比如name，不会在__getattr__中查找。
此外，注意到任意调用如s.abc都会返回None，这是因为我们定义的__getattr__默认返回就是None。要让class只响应特定的几个属性，我们就要按照约定，抛出AttributeError的错误：

In [ ]:
class Student(object):
    def __getattr__(self, attr):
        if attr=='age':
            return lambda: 25
        raise AttributeError('\'Student\' object has no attribute \'%s\'' % attr)
#这实际上可以把一个类的所有属性和方法调用全部动态化处理了，不需要任何特殊手段。

### __call__
一个对象实例可以有自己的属性和方法，当我们调用实例方法时，我们用instance.method()来调用。能不能直接在实例本身上调用呢？在Python中，答案是肯定的。
任何类，只需要定义一个__call__()方法，就可以直接对实例进行调用。请看示例：

In [16]:
class Student(object):
    def __init__(self, name):
        self.name = name

    def __call__(self):
        print('My name is %s.' % self.name)
s = Student('Michael')
s() # self参数不要传入

My name is Michael.


__call__()还可以定义参数。对实例进行直接调用就好比对一个函数进行调用一样，所以你完全可以把对象看成函数，把函数看成对象，因为这两者之间本来就没啥根本的区别。

如果你把对象看成函数，那么函数本身其实也可以在运行期动态创建出来，因为类的实例都是运行期创建出来的，这么一来，我们就模糊了对象和函数的界限。

那么，怎么判断一个变量是对象还是函数呢？其实，更多的时候，我们需要判断一个对象是否能被调用，能被调用的对象就是一个Callable对象，比如函数和我们上面定义的带有__call__()的类实例：

In [ ]:
callable(Student())       
callable(max)
#通过callable()函数，我们就可以判断一个对象是否是“可调用”对象。

### 枚举类
的枚举类型定义一个class类型，然后，每个常量都是class的一个唯一实例。Python提供了Enum类来实现这个功能：

In [24]:
from enum import Enum

Month = Enum('Month', ('Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'))
Month
print(Month.__members__.items())
#获得了Month类型的枚举类，可以直接使用Month.Jan来引用一个常量，或者枚举它的所有成员
for name, member in Month.__members__.items():
    print(name, '=>', member, ',', member.value)
    #value属性则是自动赋给成员的int常量，默认从1开始计数。



odict_items([('Jan', <Month.Jan: 1>), ('Feb', <Month.Feb: 2>), ('Mar', <Month.Mar: 3>), ('Apr', <Month.Apr: 4>), ('May', <Month.May: 5>), ('Jun', <Month.Jun: 6>), ('Jul', <Month.Jul: 7>), ('Aug', <Month.Aug: 8>), ('Sep', <Month.Sep: 9>), ('Oct', <Month.Oct: 10>), ('Nov', <Month.Nov: 11>), ('Dec', <Month.Dec: 12>)])
Jan => Month.Jan , 1
Feb => Month.Feb , 2
Mar => Month.Mar , 3
Apr => Month.Apr , 4
May => Month.May , 5
Jun => Month.Jun , 6
Jul => Month.Jul , 7
Aug => Month.Aug , 8
Sep => Month.Sep , 9
Oct => Month.Oct , 10
Nov => Month.Nov , 11
Dec => Month.Dec , 12


如果需要更精确地控制枚举类型，可以从Enum派生出自定义类：

In [25]:
from enum import Enum, unique

@unique
class Weekday(Enum):
    Sun = 0 # Sun的value被设定为0
    Mon = 1
    Tue = 2
    Wed = 3
    Thu = 4
    Fri = 5
    Sat = 6
day1 = Weekday.Mon
print(day1)
Weekday.Mon
#可见，既可以用成员名称引用枚举常量，又可以直接根据value的值获得枚举常量。
#Enum可以把一组相关常量定义在一个class中，且class不可变，而且成员可以直接比较。

Weekday.Mon


<Weekday.Mon: 1>

## 元类
### type()
动态语言和静态语言最大的不同，就是函数和类的定义，不是编译时定义的，而是运行时动态创建的。
比方说我们要定义一个Hello的class，就写一个hello.py模块：

In [28]:
class Hello(object):
    def hello(self, name='world'):
        print('Hello, %s.' % name)

当Python解释器载入hello模块时，就会依次执行该模块的所有语句，执行结果就是动态创建出一个Hello的class对象，测试如下：

In [29]:
from hello import Hello
h = Hello()
h.hello()
print(type(Hello))
print(type(h))

Hello, world.
<class 'type'>
<class 'hello.Hello'>


type()函数可以查看一个类型或变量的类型，Hello是一个class，它的类型就是type，而h是一个实例，它的类型就是class Hello。

type()函数既可以返回一个对象的类型，又可以创建出新的类型，比如，我们可以通过type()函数创建出Hello类，而无需通过class Hello(object)...的定义：

In [ ]:
def fn(self, name='world'): # 先定义函数
    print('Hello, %s.' % name)
Hello = type('Hello', (object,), dict(hello=fn)) # 创建Hello class

要创建一个class对象，type()函数依次传入3个参数：
class的名称；
继承的父类集合，注意Python支持多重继承，如果只有一个父类，别忘了tuple的单元素写法；
class的方法名称与函数绑定，这里我们把函数fn绑定到方法名hello上。

通过type()函数创建的类和直接写class是完全一样的，因为Python解释器遇到class定义时，仅仅是扫描一下class定义的语法，然后调用type()函数创建出class。
正常情况下，我们都用class Xxx...来定义类，但是，type()函数也允许我们动态创建出类来，也就是说，动态语言本身支持运行期动态创建类，这和静态语言有非常大的不同，要在静态语言运行期创建类，必须构造源代码字符串再调用编译器，或者借助一些工具生成字节码实现，本质上都是动态编译，会非常复杂。

### metaclass
除了使用type()动态创建类以外，要控制类的创建行为，还可以使用metaclass。
metaclass，直译为元类，简单的解释就是：
当我们定义了类以后，就可以根据这个类创建出实例，所以：先定义类，然后创建实例。
但是如果我们想创建出类呢？那就必须根据metaclass创建出类，所以：先定义metaclass，然后创建类。
连接起来就是：先定义metaclass，就可以创建类，最后创建实例。
所以，metaclass允许你创建类或者修改类。换句话说，你可以把类看成是metaclass创建出来的“实例”。
metaclass是Python面向对象里最难理解，也是最难使用的魔术代码。正常情况下，你不会碰到需要使用metaclass的情况，所以，以下内容看不懂也没关系，因为基本上你不会用到。

我们先看一个简单的例子，这个metaclass可以给我们自定义的MyList增加一个add方法：
定义ListMetaclass，按照默认习惯，metaclass的类名总是以Metaclass结尾，以便清楚地表示这是一个metaclass：

In [ ]:
# metaclass是类的模板，所以必须从`type`类型派生：
class ListMetaclass(type):
    def __new__(cls, name, bases, attrs):
        attrs['add'] = lambda self, value: self.append(value)
        return type.__new__(cls, name, bases, attrs)
#有了ListMetaclass，我们在定义类的时候还要指示使用ListMetaclass来定制类，传入关键字参数metaclass：    
class MyList(list, metaclass=ListMetaclass):
    pass

当我们传入关键字参数metaclass时，魔术就生效了，它指示Python解释器在创建MyList时，要通过ListMetaclass.__new__()来创建，在此，我们可以修改类的定义，比如，加上新的方法，然后，返回修改后的定义。
__new__()方法接收到的参数依次是：
1.当前准备创建的类的对象；
2.类的名字；
3.类继承的父类集合；
4.类的方法集合。

In [ ]:
#测试一下MyList是否可以调用add()方法
L = MyList()
L.add(1)
L
#而普通的list没有add()方法：
L2 = list()
L2.add(1)

# 错误，调试和测试
高级语言通常都内置了一套try...except...finally...的错误处理机制

## try

In [41]:
try:
    print('try...')
    r = 10 / 0
    print('result:', r)
except ZeroDivisionError as e:
    print('except:', e)
finally:
    print('finally...')
print('END')

try...
except: division by zero
finally...
END


从输出可以看到，当错误发生时，后续语句print('result:', r)不会被执行，except由于捕获到ZeroDivisionError，因此被执行。最后，finally语句被执行。然后，程序继续按照流程往下走。
如果把除数0改成2，则执行结果如下：

In [31]:
try:
    print('try...')
    r = 10 / 2
    print('result:', r)
except ZeroDivisionError as e:
    print('except:', e)
finally:
    print('finally...')
print('END')

try...
result: 5.0
finally...
END


错误应该有很多种类，如果发生了不同类型的错误，应该由不同的except语句块处理。没错，可以有多个except来捕获不同类型的错误：

In [ ]:
try:
    print('try...')
    r = 10 / int('a')
    print('result:', r)
except ValueError as e:
    print('ValueError:', e)
except ZeroDivisionError as e:
    print('ZeroDivisionError:', e)
finally:
    print('finally...')
print('END')

Python的错误其实也是class，所有的错误类型都继承自BaseException，所以在使用except时需要注意的是，它不但捕获该类型的错误，还把其子类也“一网打尽”。比如：

In [ ]:
try:
    foo()
except ValueError as e:
    print('ValueError')
except UnicodeError as e:
    print('UnicodeError')

第二个except永远也捕获不到UnicodeError，因为UnicodeError是ValueError的子类，如果有，也被第一个except给捕获了。
Python所有的错误都是从BaseException类派生的，常见的错误类型和继承关系看这里：
https://docs.python.org/3/library/exceptions.html#exception-hierarchy
也就是说，不需要在每个可能出错的地方去捕获错误，只要在合适的层次去捕获错误就可以了。这样一来，就大大减少了写try...except...finally的麻烦。

## 调用栈
如果错误没有被捕获，它就会一直往上抛，最后被Python解释器捕获，打印一个错误信息，然后程序退出。来看看err.py：

In [35]:
# err.py:
def foo(s):
    return 10 / int(s)

def bar(s):
    return foo(s) * 2

def main():
    bar('0')

main()

ZeroDivisionError: division by zero

$ python3 err.py
出错并不可怕，可怕的是不知道哪里出错了。解读错误信息是定位错误的关键。我们从上往下可以看到整个错误的调用函数链：


Traceback (most recent call last):           告诉我们这是错误的跟踪信息。
  File "err.py", line 11, in <module>        调用main()出错了，在代码文件err.py的   main()                                     第11行代码     
  
  File "err.py", line 9, in main             调用bar('0')出错了，在代码文件err.py   bar('0')                                   的第9行代码，但原因是第6行
    
  File "err.py", line 6, in bar              原因是return foo(s) * 2这个语句出错   return foo(s) * 2                          了，但这还不是最终原因
    
  File "err.py", line 3, in foo              原因是return 10 / int(s)这个语句出   return 10 / int(s)                        错，这是错误产生的源头，因为下面打印：
    
ZeroDivisionError: division by zero

根据错误类型ZeroDivisionError，我们判断，int(s)本身并没有出错，但是int(s)返回0，在计算10 / 0时出错，至此，找到错误源头。
出错的时候，一定要分析错误的调用栈信息，才能定位错误的位置。

### 记录错误
如果不捕获错误，自然可以让Python解释器来打印出错误堆栈，但程序也被结束了。既然我们能捕获错误，就可以把错误堆栈打印出来，然后分析错误原因，同时，让程序继续执行下去。
Python内置的logging模块可以非常容易地记录错误信息：

In [50]:
# err_logging.py

import logging

def foo(s):
    return 10 / int(s)

def bar(s):
    return foo(s) * 2

def main():
    try:
        bar('0')
    except Exception as e:
        logging.exception(e)

main()
print('END')
#同样是出错，但程序打印完错误信息后会继续执行，并正常退出

ERROR:root:division by zero
Traceback (most recent call last):
  File "<ipython-input-50-6c56206efa60>", line 13, in main
    bar('0')
  File "<ipython-input-50-6c56206efa60>", line 9, in bar
    return foo(s) * 2
  File "<ipython-input-50-6c56206efa60>", line 6, in foo
    return 10 / int(s)
ZeroDivisionError: division by zero


END


### 抛出错误
因为错误是class，捕获一个错误就是捕获到该class的一个实例。因此，错误并不是凭空产生的，而是有意创建并抛出的。Python的内置函数会抛出很多类型的错误，我们自己编写的函数也可以抛出错误。
如果要抛出错误，首先根据需要，可以定义一个错误的class，选择好继承关系，然后，用raise语句抛出一个错误的实例：

In [44]:
# err_raise.py
class FooError(ZeroDivisionError):    #继承ValueError
    pass

def foo(s):
    n = int(s)
    if n==0:
        raise FooError('invalid value -----> %s' % s)
    return 10 / n

#foo('0')
foo(0)
#只有在必要的时候才定义我们自己的错误类型。如果可以选择Python已有的内置的错误类型（比如ValueError，TypeError），
#尽量使用Python内置的错误类型。

FooError: invalid value -----> 0

捕获错误目的只是记录一下，便于后续追踪。但是，由于当前函数不知道应该怎么处理该错误，所以，最恰当的方式是继续往上抛，让顶层调用者去处理:

In [47]:
def foo(s):
    n = int(s)
    if n==0:
        raise ValueError('invalid value -------> %s' % s)
    return 10 / n

def bar():
    try:
        foo('0')
    except ValueError as e:               #捕获了错误
        print('ValueError!')              #打印一个ValueError
        raise                             #把错误通过raise语句抛出去

bar()
#raise语句如果不带参数，就会把当前错误原样抛出。

ValueError!


## 调试
一整套调试程序的手段来修复bug。

In [51]:
def foo(s):
    n = int(s)
    print('>>> n = %d' % n)
    return 10 / n

def main():
    foo('0')

main()
#用print()最大的坏处是将来还得删掉它，想想程序里到处都是print()，
#运行结果也会包含很多垃圾信息。所以，我们又有第二种方法。

>>> n = 0


ZeroDivisionError: division by zero

### 断言
凡是用print()来辅助查看的地方，都可以用断言（assert）来替代：

In [54]:
def foo(s):
    n = int(s)
    assert n != 0, 'n is zero!'
    return 10 / n

def main():
    foo('0')
main()
#,assert的意思是，表达式n != 0应该是True,断言失败，assert语句本身就会抛出AssertionError：

AssertionError: n is zero!

#启动Python解释器时可以用-O参数来关闭assert
$ python -O err.py
Traceback (most recent call last):
  ...
ZeroDivisionError: division by zero
关闭后，你可以把所有的assert语句当成pass来看。

### logging
把print()替换为logging是第3种方式，和assert比，logging不会抛出错误，而且可以输出到文件

In [63]:
import logging
logging.basicConfig(level=logging.INFO)
s = '0'
n = int(s)
#print('n = %d' % n)
logging.info('n = %d' % n)
print(10 / n)

ZeroDivisionError: division by zero

这就是logging的好处，它允许你指定记录信息的级别，有debug，info，warning，error等几个级别，当我们指定level=INFO时，logging.debug就不起作用了。同理，指定level=WARNING后，debug和info就不起作用了。这样一来，你可以放心地输出不同级别的信息，也不用删除，最后统一控制输出哪个级别的信息。

logging的另一个好处是通过简单的配置，一条语句可以同时输出到不同的地方，比如console和文件。
$ python err.py
INFO:root:n = 0
Traceback (most recent call last):
  File "err.py", line 8, in <module>
    print(10 / n)
ZeroDivisionError: division by zero

### pdb
启动Python的调试器pdb，让程序以单步方式运行，可以随时查看运行状态。

s = '0'
n = int(s)
print(10 / n)
$ python -m pdb err.py
> /Users/michael/Github/learn-python3/samples/debug/err.py(2)<module>()
-> s = '0'
以参数-m pdb启动后，pdb定位到下一步要执行的代码-> s = '0'。输入命令l来查看代码
(Pdb) n
> /Users/michael/Github/learn-python3/samples/debug/err.py(3)<module>()
-> n = int(s)
(Pdb) n
> /Users/michael/Github/learn-python3/samples/debug/err.py(4)<module>()
-> print(10 / n)
任何时候都可以输入命令p 变量名来查看变量：
(Pdb) p s
'0'
(Pdb) p n
0
(Pdb) q #输入命令q结束调试，退出程序

### pdb.set_trace()
只需要import pdb，然后，在可能出错的地方放一个pdb.set_trace()，就可以设置一个断点：

In [ ]:
# err.py
import pdb

s = '0'
n = int(s)
pdb.set_trace() # 运行到这里会自动暂停
print(10 / n)

运行代码，程序会自动在pdb.set_trace()暂停并进入pdb调试环境，可以用命令p查看变量，或者用命令c继续运行：
$ python err.py 
> /Users/michael/Github/learn-python3/samples/debug/err.py(7)<module>()
-> print(10 / n)
(Pdb) p n
0
(Pdb) c
Traceback (most recent call last):
  File "err.py", line 7, in <module>
    print(10 / n)
ZeroDivisionError: division by zero

### IDE
如果要比较爽地设置断点、单步执行，就需要一个支持调试功能的IDE。目前比较好的Python IDE有：
Visual Studio Code：https://code.visualstudio.com/，需要安装Python插件。
PyCharm：http://www.jetbrains.com/pycharm/

# IO编程
IO在计算机中指Input/Output，也就是输入和输出。由于程序和运行时数据是在内存中驻留，由CPU这个超快的计算核心来执行，涉及到数据交换的地方，通常是磁盘、网络等，就需要IO接口。
Stream（流）是一个很重要的概念，可以把流想象成一个水管，数据就是水管里的水，但是只能单向流动。Input Stream就是数据从外面（磁盘、网络）流进内存，Output Stream就是数据从内存流到外面去。对于浏览网页来说，浏览器和新浪服务器之间至少需要建立两根水管，才可以既能发数据，又能收数据。
同步IO和异步IO的区别就在于是否等待IO执行的结果，使用异步IO来编写程序性能会远远高于同步IO，异步IO的复杂度远远高于同步IO。
操作IO的能力都是由操作系统提供的，每一种编程语言都会把操作系统提供的低级C接口封装起来方便使用

## 文件读写
读写文件是最常见的IO操作。Python内置了读写文件的函数，用法和C是兼容的。在磁盘上读写文件的功能都是由操作系统提供的，现代操作系统不允许普通的程序直接操作磁盘，所以，读写文件就是请求操作系统打开一个文件对象（通常称为文件描述符），然后，通过操作系统提供的接口从这个文件对象中读取数据（读文件），或者把数据写入这个文件对象（写文件）。

### 读文件
以读文件的模式打开一个文件对象，使用Python内置的open()函数，传入文件名和标示符：

In [ ]:
f = open('/Users/michael/test.txt', 'r')#标示符'r'表示读
#如果文件不存在，open()函数就会抛出一个IOError的错误，
#并且给出错误码和详细的信息告诉你文件不存在：

In [ ]:
#如果文件打开成功，接下来，调用read()方法可以一次读取文件的全部内容，
#Python把内容读到内存，用一个str对象表示：
f.read()
#最后一步是调用close()方法关闭文件。文件使用完毕后必须关闭，
#因为文件对象会占用操作系统的资源，并且操作系统同一时间能打开的文件数量也是有限的：
f.close()

由于文件读写时都有可能产生IOError，一旦出错，后面的f.close()就不会调用。所以，为了保证无论是否出错都能正确地关闭文件，我们可以使用try ... finally来实现：

In [ ]:
try:
    f = open('/path/to/file', 'r')
    print(f.read())
finally:
    if f:
        f.close()
#等价于
#但是每次都这么写实在太繁琐，所以，Python引入了with语句来自动帮我们调用close()方法：
with open('/path/to/file', 'r') as f:
    print(f.read())

调用read()会一次性读取文件的全部内容，如果文件有10G，内存就爆了；
1）所以，要保险起见，可以反复调用read(size)方法，每次最多读取size个字节的内容；
2）另外，调用readline()可以每次读取一行内容；
3）调用readlines()一次读取所有内容并按行返回list。
如果文件很小，read()一次性读取最方便；如果不能确定文件大小，反复调用read(size)比较保险；如果是配置文件，调用readlines()最方便：

In [ ]:
for line in f.readlines():
    print(line.strip()) # 把末尾的'\n'删掉

### file-like Object
像open()函数返回的这种有个read()方法的对象，在Python中统称为file-like Object。除了file外，还可以是内存的字节流，网络流，自定义流等等。file-like Object不要求从特定类继承，只要写个read()方法就行。
StringIO就是在内存中创建的file-like Object，常用作临时缓冲。


### 二进制文件
前面讲的默认都是读取文本文件，并且是UTF-8编码的文本文件。
要读取二进制文件，比如图片、视频等等，用'rb'模式打开文件即可：

In [3]:
f = open('C:/Users/Administrator/test.jpg', 'rb')
f.read()
#windows下是\,linux和unix下是/
# \ 也是转义字符的起始字符
# 路径中的 \ 通常需要使用 \\
#C语言的文件操作，c:\\kkk\\dd.txt 和 c:/kkk/dd.txt一样

### 字符编码
要读取非UTF-8编码的文本文件，需要给open()函数传入encoding参数，例如，读取GBK编码的文件：
f = open('/Users/michael/gbk.txt', 'r', encoding='gbk')
f.read()
遇到有些编码不规范的文件，你可能会遇到UnicodeDecodeError，因为在文本文件中可能夹杂了一些非法编码的字符。遇到这种情况，open()函数还接收一个errors参数，表示如果遇到编码错误后如何处理。最简单的方式是直接忽略：
f = open('/Users/michael/gbk.txt', 'r', encoding='gbk', errors='ignore')

### 写文件
写文件和读文件是一样的，唯一区别是调用open()函数时，传入标识符'w'或者'wb'表示写文本文件或写二进制文件：
f = open('/Users/michael/test.txt', 'w')
f.write('Hello, world!')
f.close()
可以反复调用write()来写入文件，但是务必要调用f.close()来关闭文件。当我们写文件时，操作系统往往不会立刻把数据写入磁盘，而是放到内存缓存起来，空闲的时候再慢慢写入。只有调用close()方法时，操作系统才保证把没有写入的数据全部写入磁盘。忘记调用close()的后果是数据可能只写了一部分到磁盘，剩下的丢失了。所以，还是用with语句来得保险：
with open('/Users/michael/test.txt', 'w') as f:
    f.write('Hello, world!')
要写入特定编码的文本文件，请给open()函数传入encoding参数，将字符串自动转换成指定编码。
细心的童鞋会发现，以'w'模式写入文件时，如果文件已存在，会直接覆盖（相当于删掉后新写入一个文件）。如果我们希望追加到文件末尾怎么办？可以传入'a'以追加（append）模式写入。
总结：在Python中，文件读写是通过open()函数打开的文件对象完成的。使用with语句操作文件IO是个好习惯。
##### StringIO和BytesIO
是在内存中操作str和bytes的方法，使得和读写文件具有一致的接口。

## 操作文件和目录
要操作文件、目录，可以在命令行下面输入操作系统提供的各种命令来完成。比如dir、cp等命令。
如果要在Python程序中执行这些目录和文件的操作怎么办？其实操作系统提供的命令只是简单地调用了操作系统提供的接口函数，Python内置的os模块也可以直接调用操作系统提供的接口函数。
打开Python交互式命令行，我们来看看如何使用os模块的基本功能：

In [5]:
import os
os.name # 操作系统类型 
#posix，说明系统是Linux、Unix或Mac OS X；    nt，就是Windows
#os模块的某些函数是跟操作系统相关的。

'nt'

### 环境变量
在操作系统中定义的环境变量，全部保存在os.environ这个变量中，可以直接查看：

In [7]:
os.environ
os.environ.get('PATH')  #获取某个环境变量的值

'D:\\Aconda;D:\\Aconda\\Library\\mingw-w64\\bin;D:\\Aconda\\Library\\usr\\bin;D:\\Aconda\\Library\\bin;D:\\Aconda\\Scripts;D:\\Aconda;D:\\Aconda\\Scripts;D:\\PyCharm 2018.2.3\\;D:\\PyCharm 2018.2.3\\Scripts\\;D:\\Python;C:\\WINDOWS\\system32;C:\\WINDOWS;C:\\WINDOWS\\System32\\Wbem;C:\\WINDOWS\\System32\\WindowsPowerShell\\v1.0\\;C:\\WINDOWS\\System32\\OpenSSH\\;C:\\Program Files\\Microsoft VS Code\\bin;C:\\Program Files\\Git\\cmd;D:\\Aconda\\Scripts'

### 操作文件和目录
操作文件和目录的函数一部分放在os模块中，一部分放在os.path模块中

In [ ]:
# 查看当前目录的绝对路径:
os.path.abspath('.')
'/Users/michael'
# 在某个目录下创建一个新目录，首先把新目录的完整路径表示出来:
os.path.join('/Users/michael', 'testdir')
'/Users/michael/testdir'
# 然后创建一个目录:
os.mkdir('/Users/michael/testdir')
# 删掉一个目录:
os.rmdir('/Users/michael/testdir')

把两个路径合成一个时，不要直接拼字符串，而要通过os.path.join()函数，这样可以正确处理不同操作系统的路径分隔符。
而Windows下会返回这样的字符串： part-1\part-2

In [9]:
#os.path.split()函数，这样可以把一个路径拆分为两部分，
#后一部分总是最后级别的目录或文件名
os.path.split('C:/Users/Administrator/test.jpg')
#os.path.splitext()可以直接让你得到文件扩展名
os.path.splitext('C:/Users/Administrator/test.jpg')

('C:/Users/Administrator/test', '.jpg')

复制文件的函数居然在os模块中不存在
shutil模块提供了copyfile()的函数，你还可以在shutil模块中找到很多实用函数，它们可以看做是os模块的补充。

##### 利用Python的特性来过滤文件

In [11]:
#列出当前目录下的所有目录
[x for x in os.listdir('.') if os.path.isdir(x)]
#列出所有的.py文件
[x for x in os.listdir('.') if os.path.isfile(x) and os.path.splitext(x)[1]=='.py']
#Python的os模块封装了操作系统的目录和文件操作，要注意这些函数有的在os模块中，有的在os.path模块中。

['err.py', 'hello.py']

### 序列化
序列化（pickling）：把变量从内存中变成可存储或传输的对象的过程
序列化之后，就可以把序列化后的内容写入磁盘，或者通过网络传输到别的机器上。
反序列化（unpickling）：把变量内容从序列化的对象重新读到内存里
Python提供了pickle模块来实现序列化。

## JSON
如果我们要在不同的编程语言之间传递对象，就必须把对象序列化为标准格式，比如XML，但更好的方法是序列化为JSON，因为JSON表示出来就是一个字符串，可以被所有语言读取，也可以方便地存储到磁盘或者通过网络传输。JSON不仅是标准格式，并且比XML更快，而且可以直接在Web页面中读取，非常方便。
JSON表示的对象就是标准的JavaScript语言的对象，JSON和Python内置的数据类型对应如下：
JSON类型	              Python类型
  {}	                   dict
  []	                   list
 "string"	               str
 1234.56	            int或float
true/false	            True/False
  null	                   None
Python内置的json模块提供了非常完善的Python对象到JSON格式的转换。

In [1]:
#把Python对象变成一个JSON：
import json
d = dict(name='Bob', age=20, score=88)
json.dumps(d)
#dumps()方法返回一个str，内容就是标准的JSON。
#dump()方法可以直接把JSON写入一个file-like Object。

'{"name": "Bob", "age": 20, "score": 88}'

In [2]:
#把JSON反序列化为Python对象，用loads()或者对应的load()方法
#前者把JSON的字符串反序列化，后者从file-like Object中读取字符串并反序列化：
json_str = '{"age": 20, "score": 88, "name": "Bob"}'
json.loads(json_str)
#JSON标准规定JSON编码是UTF-8

{'age': 20, 'score': 88, 'name': 'Bob'}

### JSON进阶
Python的dict对象可以直接序列化为JSON的{}，不过，普通的对象不是一个可序列化为JSON的对象。
dumps()方法的参数列表，可以发现，除了第一个必须的obj参数外，dumps()方法还提供了一大堆的可选参数，可选参数就是让我们来定制JSON序列化。

In [7]:
#可选参数default就是把任意一个对象变成一个可序列为JSON的对象，
#我们只需要为Student专门写一个转换函数，再把函数传进去即可：
class Student(object):
    def __init__(self, name, age, score):
        self.name = name
        self.age = age
        self.score = score
s = Student('Bob', 20, 88)
def student2dict(std): #只需要为Student专门写一个转换函数，再把函数传进去即可
    return {
        'name': std.name,
        'age': std.age,
        'score': std.score
    }
#json.dumps(s, default=student2dict)
json.dumps(s, default=lambda obj: obj.__dict__)
#通常class的实例都有一个__dict__属性，它就是一个dict，用来存储实例变量。
#类的属性为key，属性值为value，变为字典
#也有少数例外，比如定义了__slots__的class。

'{"name": "Bob", "age": 20, "score": 88}'

In [ ]:
#把JSON反序列化为一个Student对象实例，loads()方法首先转换出一个dict对象，
#然后，我们传入的object_hook函数负责把dict转换为Student实例
def dict2student(d):
    return Student(d['name'], d['age'], d['score'])
json_str = '{"age": 20, "score": 88, "name": "Bob"}'
json.loads(json_str, object_hook=dict2student)

In [10]:
#对中文进行JSON序列化时，json.dumps()提供了一个ensure_ascii参数
obj = dict(name='小明', age=20)
s = json.dumps(obj, ensure_ascii=False)
s

'{"name": "小明", "age": 20}'

# 进程和线程
对于操作系统来说，一个任务就是一个进程（Process）
在一个进程内部，要同时干多件事，就需要同时运行多个“子任务”，我们把进程内的这些“子任务”称为线程（Thread）。
前面编写的所有的Python程序，都是执行单任务的进程，也就是只有一个线程。多任务解决：
1）多进程模式 multiprocessing
1）多线程模式
1）多进程+多线程模式
#### multiprocessing

In [ ]:
#multiprocessing模块提供了一个Process类来代表一个进程对象
from multiprocessing import Process
import os


def run_proc(name):            # 子进程要执行的代码
    print('Run child process %s (%s)...' % (name, os.getpid()))

if __name__=='__main__':
    print('Parent process %s.' % os.getpid())  #父进程
    #创建子进程时，只需要传入一个执行函数和函数的参数，创建一个Process实例
    p = Process(target=run_proc, args=('test',))
    print('Child process will start.')
    p.start()   #用start()方法启动
    p.join()    #join()方法可以等待子进程结束后再继续往下运行，通常用于进程间的同步。
    print('Child process end.')

### Pool
如果要启动大量的子进程，可以用进程池的方式批量创建子进程：

In [ ]:
from multiprocessing import Pool
import os, time, random

def long_time_task(name):           # 子进程要执行的代码
    print('Run task %s (%s)...' % (name, os.getpid()))
    start = time.time()
    time.sleep(random.random() * 3)
    end = time.time()
    print('Task %s runs %0.2f seconds.' % (name, (end - start)))

if __name__=='__main__':
    print('Parent process %s.' % os.getpid())
    p = Pool(4)
    for i in range(5):              #批量创建子进程
        p.apply_async(long_time_task, args=(i,))
    print('Waiting for all subprocesses done...')
    p.close()    #调用join()之前必须先调用close()，调用close()之后就不能继续添加新的Process了
    p.join()     #对Pool对象调用join()方法会等待所有子进程执行完毕
    print('All subprocesses done.')
#Pool的默认大小是CPU的核数

### 子进程
子进程并不是自身，而是一个外部进程。我们创建了子进程后，还需要控制子进程的输入和输出。
subprocess模块可以让我们非常方便地启动一个子进程，然后控制其输入和输出。

In [13]:
import subprocess
print('$ nslookup www.python.org')
r = subprocess.call(['nslookup', 'www.python.org'])
print(r)
print('Exit code:', r)

$ nslookup www.python.org
0
Exit code: 0


在Unix/Linux下，可以使用fork()调用实现多进程。
要实现跨平台的多进程，可以使用multiprocessing模块。

# 正则表达式
用一种描述性的语言来给字符串定义一个规则，凡是符合规则的字符串，我们就认为它“匹配”了，否则，该字符串就是不合法的。
\d     一个数字
\w     一个字母或数字
.      一个任意字符
*      任意个字符（包括0个）
+      至少一个字符
?       0个或1个字符
{n}     表示n个字符
{n,m}   n-m个字符
### 进阶
[]      表示范围
A|B     配A或B
^       表示行的开头，^\d表示必须以数字开头
$       表示行的结束，\d$表示必须以数字结束。

## re模块
Python提供re模块，包含所有正则表达式的功能。Python的字符串本身也用\转义
使用Python的r前缀，就不用考虑转义的问题

In [16]:
s = r'ABC\-001' # Python的字符串
# 对应的正则表达式字符串不变：
# 'ABC\-001'      #等价于s='ABC\\-001'
s

'ABC\\-001'

In [2]:
#判断正则表达式是否匹配：
import re
re.match(r'^\d{3}\-\d{3,8}$', '010-12345')
re.match(r'^\d{3}\-\d{3,8}$', '010 12345')
#match()方法判断是否匹配，如果匹配成功，返回一个Match对象，否则返回None。

#常用的判断方法
test = '用户输入的字符串'
if re.match(r'正则表达式', test):
    print('ok')
else:
    print('failed')

### 切分字符串

In [ ]:
'a b   c'.split(' ') #无法识别连续的空格，用正则表达式试试
re.split(r'\s+', 'a b   c')  #正则r'\s+'表示任意多个空格
re.split(r'[\s\,]+', 'a,b, c  d') #加入逗号
re.split(r'[\s\,\;]+', 'a,b;; c  d') #再加入分号

### 分组
除了简单地判断是否匹配之外，正则表达式还有提取子串的强大功能。用()表示的就是要提取的分组（Group）。

In [ ]:
#^(\d{3})-(\d{3,8})$分别定义了两个组，可以直接从匹配的字符串中提取出区号和本地号码
m = re.match(r'^(\d{3})-(\d{3,8})$', '010-12345')
m
#如果正则表达式中定义了组，就可以在Match对象上用group()方法提取出子串来。
m.group(0)
m.group(1)
m.group(2)

### 贪婪匹配
正则匹配默认是贪婪匹配，也就是匹配尽可能多的字符。

In [3]:
re.match(r'^(\d+)(0*)$', '102300').groups()
#必须让\d+采用非贪婪匹配（也就是尽可能少匹配），才能把后面的0匹配出来，加个?就可以让\d+采用非贪婪匹配：

('102300', '')

### 编译
当我们在Python中使用正则表达式时，re模块内部会干两件事情：
编译正则表达式，如果正则表达式的字符串本身不合法，会报错；
用编译后的正则表达式去匹配字符串。
如果一个正则表达式要重复使用几千次，出于效率的考虑，我们可以预编译该正则表达式，接下来重复使用时就不需要编译这个步骤了，直接匹配：

In [4]:
import re
# 编译:
#编译后生成Regular Expression对象，
#由于该对象自己包含了正则表达式，所以调用对应的方法时不用给出正则字符串
re_telephone = re.compile(r'^(\d{3})-(\d{3,8})$')
# 使用：
re_telephone.match('010-12345').groups()
re_telephone.match('010-8086').groups()

('010', '12345')

## 常用内建模块
### namedtuple
namedtuple是一个函数，它用来创建一个自定义的tuple对象，并且规定了tuple元素的个数，并可以用属性而不是索引来引用tuple的某个元素。
用namedtuple可以很方便地定义一种数据类型，它具备tuple的不变性，又可以根据属性来引用
Point对象是tuple的一种子类

In [6]:
from collections import namedtuple
# namedtuple('名称', [属性list]):
#要用坐标和半径表示一个圆
Circle = namedtuple('Circle', ['x', 'y', 'r'])
c=Circle(1,0,4)
c.r

4

### defaultdict
使用dict时，如果引用的Key不存在，就会抛出KeyError。如果希望key不存在时，返回一个默认值，就可以用defaultdict：
默认值是调用函数返回的，而函数在创建defaultdict对象时传入。
除了在Key不存在时返回默认值，defaultdict的其他行为跟dict是完全一样的。

In [ ]:
from collections import defaultdict
dd = defaultdict(lambda: 'N/A')
dd['key1'] = 'abc'
dd['key1'] # key1存在
dd['key2'] # key2不存在，返回默认值

## itertools
itertools模块提供的全部是处理迭代功能的函数，它们的返回值不是list，而是Iterator，只有用for循环迭代的时候才真正计算。
### groupby()
groupby()把迭代器中相邻的重复元素挑出来放在一起

In [8]:
import itertools
for key, group in itertools.groupby('AAABBBCCAAA'):
    print(key, list(group))

A ['A', 'A', 'A']
B ['B', 'B', 'B']
C ['C', 'C']
A ['A', 'A', 'A']


## urllib
urllib提供了一系列用于操作URL的功能。
### Get
urllib的request模块可以非常方便地抓取URL内容，也就是发送一个GET请求到指定的页面，然后返回HTTP的响应：

想模拟浏览器发送GET请求，就需要使用Request对象，通过往Request对象添加HTTP头，我们就可以把请求伪装成浏览器

In [ ]:
#例如，模拟iPhone 6去请求豆瓣首页：
from urllib import request

req = request.Request('http://www.douban.com/')
req.add_header('User-Agent', 'Mozilla/6.0 (iPhone; CPU iPhone OS 8_0 like Mac OS X) AppleWebKit/536.26 (KHTML, like Gecko) Version/8.0 Mobile/10A5376e Safari/8536.25')
with request.urlopen(req) as f:
    print('Status:', f.status, f.reason)   
    for k, v in f.getheaders():
        print('%s: %s' % (k, v))
    print('Data:', f.read().decode('utf-8'))  
#这样豆瓣会返回适合iPhone的移动版网页：

### Post
要以POST发送一个请求，只需要把参数data以bytes形式传入。

In [ ]:
#模拟一个微博登录，先读取登录的邮箱和口令，
#然后按照weibo.cn的登录页的格式以username=xxx&password=xxx的编码传入：
from urllib import request, parse

print('Login to weibo.cn...')
email = input('Email: ')
passwd = input('Password: ')
login_data = parse.urlencode([
    ('username', email),
    ('password', passwd),
    ('entry', 'mweibo'),
    ('client_id', ''),
    ('savestate', '1'),
    ('ec', ''),
    ('pagerefer', 'https://passport.weibo.cn/signin/welcome?entry=mweibo&r=http%3A%2F%2Fm.weibo.cn%2F')
])

req = request.Request('https://passport.weibo.cn/sso/login')
req.add_header('Origin', 'https://passport.weibo.cn')
req.add_header('User-Agent', 'Mozilla/6.0 (iPhone; CPU iPhone OS 8_0 like Mac OS X) AppleWebKit/536.26 (KHTML, like Gecko) Version/8.0 Mobile/10A5376e Safari/8536.25')
req.add_header('Referer', 'https://passport.weibo.cn/signin/login?entry=mweibo&res=wel&wm=3349&r=http%3A%2F%2Fm.weibo.cn%2F')

with request.urlopen(req, data=login_data.encode('utf-8')) as f:
    print('Status:', f.status, f.reason)
    for k, v in f.getheaders():
        print('%s: %s' % (k, v))
    print('Data:', f.read().decode('utf-8'))

### HTMLParser
HTML本质上是XML的子集，但是HTML的语法没有XML那么严格，所以不能用标准的DOM或SAX来解析HTML，Python提供了HTMLParser来非常方便地解析HTML

In [ ]:
from html.parser import HTMLParser
from html.entities import name2codepoint

class MyHTMLParser(HTMLParser):

    def handle_starttag(self, tag, attrs):
        print('<%s>' % tag)

    def handle_endtag(self, tag):
        print('</%s>' % tag)

    def handle_startendtag(self, tag, attrs):
        print('<%s/>' % tag)

    def handle_data(self, data):
        print(data)

    def handle_comment(self, data):
        print('<!--', data, '-->')

    def handle_entityref(self, name):
        print('&%s;' % name)

    def handle_charref(self, name):
        print('&#%s;' % name)

parser = MyHTMLParser()
parser.feed('''<html>
<head></head>
<body>
<!-- test html parser -->
    <p>Some <a href=\"#\">html</a> HTML&nbsp;tutorial...<br>END</p>
</body></html>''')

## 常用第三方库
### psutil
process and system utilities\
可以通过一两行代码实现系统监控，还可以跨平台使用，支持Linux／UNIX／OSX／Windows等
如果安装了Anaconda，psutil就已经可用了。

In [11]:
#获取CPU信息
import psutil
psutil.cpu_count() # CPU逻辑数量
psutil.cpu_count(logical=False) # CPU物理核心
#统计CPU的用户／系统／空闲时间：
psutil.cpu_times()
#使用psutil获取物理内存和交换内存信息
psutil.virtual_memory()
psutil.swap_memory()
#通过psutil获取磁盘分区、磁盘使用率和磁盘IO信息：
psutil.disk_partitions() # 磁盘分区信息
psutil.disk_usage('/') # 磁盘使用情况
psutil.disk_io_counters() # 磁盘IO
#psutil可以获取网络接口和网络连接信息：
#通过psutil可以获取到所有进程的详细信息：
psutil.pids() # 所有进程ID
p = psutil.Process(3776) # 获取指定进程ID=3776，其实就是当前Python交互环境
p.name() # 进程名称
p.exe() # 进程exe路径
p.cwd() # 进程工作目录
p.cmdline() # 进程启动的命令行
p.ppid() # 父进程ID
p.parent() # 父进程
p.children() # 子进程列表
p.status() # 进程状态
p.username() # 进程用户名
p.create_time() # 进程创建时间
p.terminal() # 进程终端
p.threads() # 所有线程信息
p.environ() # 进程环境变量
#和获取网络连接类似，获取一个root用户的进程需要root权限，
#启动Python交互环境或者.py文件时，需要sudo权限。

### virtualenv
virtualenv用来为一个应用创建一套“隔离”的Python运行环境
$ pip3 install virtualenv
假定我们要开发一个新的项目，需要一套独立的Python运行环境，可以这么做：
1）创建目录
2）创建一个独立的Python运行环境，命名为venv：
3）进入该环境
4）正常安装各种第三方包，并运行python命令
5）退出当前的venv环境
在venv环境下，用pip安装的包都被安装到venv这个环境下，系统Python环境不受任何影响。
virtualenv是如何创建“独立”的Python运行环境的呢？原理很简单，就是把系统Python复制一份到virtualenv的环境，用命令source venv/bin/activate进入一个virtualenv环境时，virtualenv会修改相关环境变量，让命令python和pip均指向当前的virtualenv环境。

### 图形界面
Python支持多种图形界面的第三方库，包括：Tk，wxWidgets，Qt，GTK等等。
但是Python自带的库是支持Tk的Tkinter，使用Tkinter，无需安装任何包，就可以直接使用。

In [12]:
from tkinter import *
import tkinter.messagebox as messagebox
#在GUI中，每个Button、Label、输入框等，都是一个Widget。
#Frame则是可以容纳其他Widget的Widget，所有的Widget组合起来就是一棵树。
class Application(Frame): #从Frame派生一个Application类，这是所有Widget的父容器
    def __init__(self, master=None):
        Frame.__init__(self, master)
        self.pack()
        self.createWidgets()

    def createWidgets(self):
        self.nameInput = Entry(self)
#pack()方法把Widget加入到父容器中，
#并实现布局。pack()是最简单的布局，grid()可以实现更复杂的布局。
        self.nameInput.pack()
#在createWidgets()方法中，
#我们创建一个Label和一个Button，当Button被点击时，触发self.quit()使程序退出。
        self.alertButton = Button(self, text='Hello', command=self.hello)
        self.alertButton.pack()

    def hello(self):
        name = self.nameInput.get() or 'world'
        messagebox.showinfo('Message', 'Hello, %s' % name)

app = Application()
# 设置窗口标题:
app.master.title('Hello World')
# 主消息循环:
app.mainloop()

# 访问数据库
程序运行的时候，数据都是在内存中的。当程序终止的时候，通常都需要将数据保存到磁盘上，无论是保存到本地磁盘，还是通过网络保存到服务器上，最终都会将数据写入磁盘文件。
而如何定义数据的存储格式就是一个大问题。
存储和读取需要自己实现，JSON还是标准，自己定义的格式就各式各样了；
不能做快速查询，只有把数据全部读到内存中才能自己遍历，但有时候数据的大小远远超过了内存（比如蓝光电影，40GB的数据），根本无法全部读入内存。
为了便于程序保存和读取数据，而且，能直接通过条件快速查询到指定的数据，就出现了数据库（Database）这种专门用于集中存储和查询的软件。
现在广泛使用的关系数据库是20世纪70年代基于关系模型的基础上诞生的。
#### 数据库类别
付费的商用数据库：
Oracle，典型的高富帅；
SQL Server，微软自家产品，Windows定制专款；
DB2，IBM的产品，听起来挺高端；
Sybase，曾经跟微软是好基友，后来关系破裂，现在家境惨淡。
Google、Facebook，还是国内的BAT，无一例外都选择了免费的开源数据库：
MySQL，大家都在用，一般错不了；
PostgreSQL，学术气息有点重，其实挺不错，但知名度没有MySQL高；
sqlite，嵌入式数据库，适合桌面和移动应用。
因为MySQL普及率最高，出了错，可以很容易找到解决方法。而且，围绕MySQL有一大堆监控和运维的工具，安装和使用很方便。
从MySQL官方网站下载并安装MySQL Community Server 5.6，这个版本是免费的
https://dev.mysql.com/downloads/mysql/
### 几个概念
表是数据库中存放关系数据的集合，一个数据库里面通常都包含多个表，比如学生的表，班级的表，学校的表，等等。表和表之间通过外键关联。
操作关系数据库，首先需要连接到数据库，一个数据库连接称为Connection；
连接到数据库后，需要打开游标，称之为Cursor，通过Cursor执行SQL语句，然后，获得执行结果；
Python定义了一套操作数据库的API接口，任何数据库要连接到Python，只需要提供符合Python标准的数据库驱动即可。

### SQLite
一种嵌入式数据库，它的数据库就是一个文件。由于SQLite本身是C写的，而且体积很小，所以，经常被集成到各种应用程序中，甚至在iOS和Android的App中都可以集成。
Python就内置了SQLite3，所以，在Python中使用SQLite，不需要安装任何东西，直接使用。

In [ ]:
# 导入SQLite驱动:
import sqlite3
# 连接到SQLite数据库
# 数据库文件是test.db
# 如果文件不存在，会自动在当前目录创建:
conn = sqlite3.connect('test.db')     #connection对象
cursor = conn.cursor()                #consor对象
cursor.execute('create table user (id varchar(20) primary key, name varchar(20))')
# 继续执行一条SQL语句，插入一条记录:
cursor.execute('insert into user (id, name) values (\'1\', \'Michael\')')
# 通过rowcount获得插入的行数:
cursor.rowcount
# 关闭Cursor:
cursor.close()
# 提交事务:
conn.commit()
# 关闭Connection:
conn.close()

使用Python的DB-API时，只要搞清楚Connection和Cursor对象，打开后一定记得关闭，就可以放心地使用。
使用Cursor对象执行insert，update，delete语句时，执行结果由rowcount返回影响的行数，就可以拿到执行结果。
使用Cursor对象执行select语句时，通过featchall()可以拿到结果集。结果集是一个list，每个元素都是一个tuple，对应一行记录。
如果SQL语句带有参数，那么需要把参数按照位置传递给execute()方法，有几个?占位符就必须对应几个参数，例如：
cursor.execute('select * from user where name=? and pwd=?', ('abc', 'password'))

#### MySQL
MySQL是Web世界中使用最广泛的数据库服务器。SQLite的特点是轻量级、可嵌入，但不能承受高并发访问，适合桌面和移动应用。而MySQL是为服务器端设计的数据库，能承受高并发访问，同时占用的内存也远远大于SQLite。
##### 安装MySQL
##### 安装MySQL驱动
由于MySQL服务器以独立的进程运行，并通过网络对外服务，所以，需要支持Python的MySQL驱动来连接到MySQL服务器。MySQL官方提供了mysql-connector-python驱动，但是安装的时候需要给pip命令加上参数--allow-external：
$ pip install mysql-connector-python --allow-external mysql-connector-python
$ pip install mysql-connector

In [ ]:
# 导入MySQL驱动:
import mysql.connector
# 注意把password设为你的root口令:
conn = mysql.connector.connect(user='root', password='password', database='test')
cursor = conn.cursor()
# 创建user表:
cursor.execute('create table user (id varchar(20) primary key, name varchar(20))')
# 插入一行记录，注意MySQL的占位符是%s:
cursor.execute('insert into user (id, name) values (%s, %s)', ['1', 'Michael'])
cursor.rowcount

# 提交事务:
conn.commit()
cursor.close()
# 运行查询:
cursor = conn.cursor()
cursor.execute('select * from user where id = %s', ('1',))
values = cursor.fetchall()
values
# 关闭Cursor和Connection:
cursor.close()
True
conn.close()
#由于Python的DB-API定义都是通用的，所以，操作MySQL的数据库代码和SQLite类似。
#执行INSERT等操作后要调用commit()提交事务；